File: cpp/src/XyzTimer.cpp
    

#include <iostream>
#include "XyzTimer.h"
#include "SessionModificationCmd.h"

XyzTimer::XyzTimer() :
        timerValue_(0), timerUnit_(XyzTimerUnit::TimerDeactivated) {}

void XyzTimer::set(XyzTimerUnit unit, int value)
{
    timerUnit_  = unit;
    timerValue_ = value;
}

void XyzTimer::encode(ByteBuffer& data)
{
    data.append(YEW::YewXyzTimer);
    uint8_t temp;
    temp = timerUnit_ << 5;

    if (timerUnit_ != XyzTimerUnit::TimerDeactivated)
    {
        if (timerValue_ <= 0x1F) // 5 bits max range
        {
            temp |= (timerValue_ & 0x1F);
        }
        else
        {
            std::cout << "XYZ timer value out of range. Encoding max value" << std::endl;
            temp |= 0x1F;
        }
    }
    data.append(temp);
}

File: java/src/main/java/com/bacheconsulting/kata/encode/XyzTimer.java
package com.bacheconsulting.kata.encode;

final class XyzTimer {

    private XyzTimerUnit timerUnit;
    private int timerValue;

    public void set(XyzTimerUnit unit, int timerValue) {
        this.timerUnit = unit;
        this.timerValue = timerValue;
    }

    public void encode(ByteBuffer data) {
        data.append(YEW.YEW_XYZ_TIMER.code());
        int temp;
        temp = (this.timerUnit.ordinal() << 5);

        if (this.timerUnit != XyzTimerUnit.TIMER_DEACTIVATED) {
            if (this.timerValue <= 0x1F) {
                temp |= (this.timerValue & 0x1F);
            } else {
                System.out.print("XYZ timer value out of range. Encoding max value\n");
                temp |= 0x1F;
            }
        }
        data.append(temp);
    }
}